File: Python/325__Maximum_Size_Subarray_Sum_Equals_K.py
class Solution(object):
    def maxSubArrayLen(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: int
        """
        s = 0
        dic = {}
        result = 0
        for i, num in enumerate(nums):
            s += num
            if s == k:
                result = max(result, i + 1)
            elif s - k in dic:
                result = max(result, i - dic[s - k])
            if s not in dic:
                dic[s] = i
        return result

File: Java/325__Maximum_Size_Subarray_Sum_Equals_k.java
class Solution {
    public int maxSubArrayLen(int[] nums, int k) {
        int cur_sum = 0, result = 0;
        HashMap<Integer, Integer> map = new HashMap<>();
        map.put(0, -1);
        for (int i = 0; i < nums.length; i++) {
            cur_sum += nums[i];
            if (!map.containsKey(cur_sum)) {
                map.put(cur_sum, i);
            }
            int previousSum = cur_sum - k;
            if (map.containsKey(previousSum)) {
                int length = i - map.get(previousSum);
                result = Math.max(length, result);
            }
        }

        return result;
    }
}
